src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -214,46 +214,26 @@ void GcsPlacementGroupScheduler::ScheduleUnplacedBundles(
     return;
   }
 
-  // If schedule success, the decision will be set as schedule_map[bundles[pos]]
-  // else will be set ClientID::Nil().
-  auto bundle_locations = std::make_shared<BundleLocations>();
-  // To count how many scheduler have been return, which include success and failure.
-  auto finished_count = std::make_shared<size_t>();
-  RAY_CHECK(
-      placement_group_leasing_in_progress_.emplace(placement_group->GetPlacementGroupID())
-          .second);
+  auto leasing_context = std::make_shared<LeasingContext>(placement_group, bundles);
+  RAY_CHECK(placement_group_leasing_in_progress_
+                .emplace(placement_group->GetPlacementGroupID(), leasing_context)
+                .second);
 
   /// TODO(AlisaWu): Change the strategy when reserve resource failed.
   for (auto &bundle : bundles) {
     const auto &bundle_id = bundle->BundleId();
     const auto &node_id = selected_nodes[bundle_id];
-    RAY_CHECK(node_to_bundles_when_leasing_[node_id].emplace(bundle_id).second);
-
-    ReserveResourceFromNode(
-        bundle, gcs_node_manager_.GetNode(node_id),
-        [this, bundle_id, bundle, bundles, node_id, placement_group, bundle_locations,
-         finished_count, failure_callback, success_callback](const Status &status) {
-          auto leasing_bundles = node_to_bundles_when_leasing_.find(node_id);
-          RAY_CHECK(leasing_bundles != node_to_bundles_when_leasing_.end());
-          auto bundle_iter = leasing_bundles->second.find(bundle->BundleId());
-          RAY_CHECK(bundle_iter != leasing_bundles->second.end());
-          // Remove the bundle from the leasing map as the reply is returned from the
-          // remote node.
-          leasing_bundles->second.erase(bundle_iter);
-          if (leasing_bundles->second.empty()) {
-            node_to_bundles_when_leasing_.erase(leasing_bundles);
-          }
-
-          if (status.ok()) {
-            (*bundle_locations)[bundle_id] = std::make_pair(node_id, bundle);
-          }
-
-          if (++(*finished_count) == bundles.size()) {
-            OnAllBundleSchedulingRequestReturned(placement_group, bundles,
-                                                 bundle_locations, failure_callback,
-                                                 success_callback);
-          }
-        });
+    leasing_context->MarkLeaseStarted(node_id, bundle);
+
+    ReserveResourceFromNode(bundle, gcs_node_manager_.GetNode(node_id),
```

Comment:
@rkooo567 I'm not sure if callback will be called if raylet is dead, do you know the logic of this?

Reply:
I think it wont. Are you suggesting to have a callback map and empty them when the node is dead or placement group is removed?

Reply:
I will address this in the next PR. Added a comment for now.

Reply:
Added a comment for now.